cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java

```diff
@@ -283,23 +342,27 @@ public void drop(Id.DatasetInstance instance) throws Exception {
    * @throws NamespaceNotFoundException if the requested namespace was not found
    * @throws IOException if there was a problem in checking if the namespace exists over HTTP
    */
-  public DatasetAdminOpResponse executeAdmin(Id.DatasetInstance instance, String method) throws Exception {
+  DatasetAdminOpResponse executeAdmin(Id.DatasetInstance instance, String method) throws Exception {
     // Throws NamespaceNotFoundException if the namespace does not exist
     ensureNamespaceExists(instance.getNamespace());
 
     Object result = null;
 
     // NOTE: one cannot directly call create and drop, instead this should be called thru
     //       POST/DELETE @ /data/datasets/{instance-id}. Because we must create/drop metadata for these at same time
+    Principal principal = SecurityRequestContext.toPrincipal();
     switch (method) {
       case "exists":
+        ensureAccess(instance.toEntityId());
         result = opExecutorClient.exists(instance);
         break;
       case "truncate":
+        authorizationEnforcementService.enforce(instance.toEntityId(), principal, Action.ADMIN);
         opExecutorClient.truncate(instance);
         publishAudit(instance, AuditType.TRUNCATE);
         break;
       case "upgrade":
+        authorizationEnforcementService.enforce(instance.toEntityId(), principal, Action.ADMIN);
```

Comment:
Either that, or they have to be a super user. Which is a setting that authorization extensions can honor - e.g. the cdap-sentry-extension accepts a list of super users. Is that fair? I wanted to add this here, because even though right now this is called from the upgrade tool on all datasets at once, there may be ways to upgrade a single dataset in future, and this seemed like an admin operation on a dataset.